src/toast/ops/demodulation.py

```diff
@@ -0,0 +1,701 @@
+            raise RuntimeError(
+                "None of the observations have a spinning HWP.  Nothing to demodulate."
+            )
+
+        # Each modulated detector demodulates into 3 or 5 pseudo detectors
+
+        self.prefixes = ["demod0", "demod4r", "demod4i"]
+        if self.do_2f:
+            self.prefixes.extend(["demod2r", "demod2i"])
+
+        for obs in demodulate_obs:
+            dets = obs.select_local_detectors(detectors)
+
+            offset = obs.local_index_offset
+            nsample = obs.n_local_samples
+
+            fsample = obs.telescope.focalplane.sample_rate
+            fmax, hwp_rate = self._get_fmax(obs)
+            wkernel = self._get_wkernel(fmax, fsample)
+            lowpass = Lowpass(wkernel, fmax, fsample, offset, self.nskip, self.window)
+
+            # Create a new observation to hold the demodulated and downsampled data
+
+            demod_telescope = self._demodulate_telescope(obs)
+            demod_times = self._demodulate_times(obs)
+            demod_detsets = self._demodulate_detsets(obs)
+            demod_sample_sets = self._demodulate_sample_sets(obs)
+            demod_process_rows = obs.dist.process_rows
+
+            demod_name = f"demod_{obs.name}"
+            demod_obs = Observation(
+                demod_telescope,
+                demod_times.size,
+                name=demod_name,
+                uid=name_UID(demod_name),
+                comm=obs.comm,
+                detector_sets=demod_detsets,
+                process_rows=demod_process_rows,
+                sample_sets=demod_sample_sets,
+            )
+
+            # Allocate storage
+
+            demod_dets = []
+            for det in dets:
+                for prefix in self.prefixes:
+                    demod_dets.append(f"{prefix}_{det}")
+            n_local = demod_obs.n_local_samples
+
+            demod_obs.shared.create(self.times, (n_local,), comm=demod_obs.comm_col)
+            demod_obs.shared[self.times].set(demod_times, offset=(0,), fromrank=0)
+            demod_obs.shared.create(
+                self.boresight, (n_local, 4), comm=demod_obs.comm_col
+            )
+            demod_obs.shared.create(
+                self.shared_flags, (n_local,), dtype=np.uint8, comm=demod_obs.comm_col
+            )
+
+            demod_obs.detdata.ensure(
+                self.det_data, detectors=demod_dets, dtype=np.float64
+            )
+            demod_obs.detdata.ensure(
+                self.det_flags, detectors=demod_dets, dtype=np.uint
+            )
+
+            self._demodulate_flags(obs, demod_obs, dets, wkernel, offset)
+            self._demodulate_signal(data, obs, demod_obs, dets, lowpass)
+            self._demodulate_pointing(data, obs, demod_obs, dets, lowpass, offset)
+            self._demodulate_noise(obs, demod_obs, dets, fsample, hwp_rate, lowpass)
+
+            self._demodulate_intervals(obs, demod_obs)
+
+            self.demod_data.obs.append(demod_obs)
```

Comment:
What happens if there are other shared or detdata objects needed downstream for processing the demodulated data or other studies?  Will we add those one at a time to this operator as needed?  Should we provide List traits with the names of additional shared or detdata objects we want downsampled a propagated to the output data?

Reply:
Different data objects like pointing, flags and signal require separate demodulation methods. Supporting additional flavors would require providing a list for each flavor. I would hold off adding that capability until there is a demonstrated need.